File: src/amqp_director.erl
-module (amqp_director).
-compile([{parse_transform, lager_transform}]).

-include_lib("amqp_client/include/amqp_client.hrl").
-deprecated({client_child_spec, 3, eventually}).

%% Lifetime API
-export([
         ad_client_child_spec/3,
         sp_client_child_spec/3,
         client_child_spec/3,
         mk_app_id/1,
         parse_connection_parameters/1,
         server_child_spec/5
        ]).


-type connection_option() ::
    {host, binary()} |
    {port, non_neg_integer()} |
    {username, binary()} |
    {password, binary()} |
    {virtual_host, binary()}.

-type parsed_connection_options() :: #amqp_params_network{}.

-type content_type() :: binary().
-type queue_definition() :: #'queue.declare'{} | #'queue.bind'{} | #'exchange.declare'{}.
-type handler_return_type() ::  {reply, Payload :: binary(), content_type()} |
                                reject |
                                reject_no_requeue |
                                {reject_dump_msg, binary()} |
                                ack.
-type handler() :: fun((Payload :: binary(), ContentType :: binary(), Type :: binary()) -> handler_return_type()).

-type server_option() ::
    {queue_definitions, list(queue_definition())} |
    {consume_queue, binary()} |
    {consumer_tag, binary()} |
    {no_ack, boolean()} |
    {qos, number()} |
    {reply_persistent, boolean()}.


-type client_option() ::
    {app_id, binary()} |
    {queue_definitions, list(queue_definition())} |
    {reply_queue, binary()} |
    {no_ack, boolean()}.


-type pull_client_option() ::
    {app_id, binary()} |
    {queue_definitions, list(queue_definition())}.



%% @doc Construct an application Id for this node based on a RegName atom
%% @end
-spec mk_app_id(atom()) -> binary().
mk_app_id(RegName) when is_atom(RegName) ->
  Hostname = string:strip(os:cmd("/bin/hostname"), right, $\n),
  Creation = erlang:system_info(creation),
  {Mega, S, _} = os:timestamp(),
  iolist_to_binary(
    [Hostname, $-, atom_to_list(node()), $-,
     integer_to_list(Creation), $.,
     integer_to_list(Mega * 1000000 + S), $., atom_to_list(RegName)]).

%% @doc
%% Construct a child specification for an AMQP RPC Server.
%% This specification allows for RPC servers to be nested under any supervisor in the application using AmqpDirector. The RPC Server
%% will initialize the queues it is instructed to and will then consume messages on the queue specified. The handler function will
%% be called to handle each request.
%% @end
-spec server_child_spec(atom(), handler(), parsed_connection_options(), pos_integer(), list(server_option())) -> supervisor:child_spec().
server_child_spec(Name, Fun, ConnInfo, ServersCount, Config) ->
	ConnReg = list_to_atom(atom_to_list(Name) ++ "_conn"),
    {Name, {amqp_server_sup, start_link, [ConnReg, ConnInfo, Config, Fun, ServersCount]},
     permanent, infinity, supervisor, [amqp_server_sup]}.

%% @doc
%% Construct a child specification for an AMQP RPC Server.
%% This specification allows for RPC clients to be nested under any supervisor in the application using AmqpDirector. The RPC
%% client can perform queue initialization. It will also create a reply queue to consume replies on.
%% @end
-spec ad_client_child_spec(atom(), parsed_connection_options(), list(client_option())) -> supervisor:child_spec().
ad_client_child_spec(Name, ConnInfo, Config) ->
    ConnReg = list_to_atom(atom_to_list(Name) ++ "_conn"),
    {Name, {amqp_client_sup, start_link_ad, [Name, ConnReg, ConnInfo, Config]},
     permanent, infinity, supervisor, [amqp_client_sup]}.

%% @doc
%% Creates a child specification for an AMQP RPC pull client.
%% This specification allows for RPC clients to be nested under any supervisor in the application using AmqpDirector. The pull client
%% uses the Synchronous Pull (`#basic.get{}') over AMQP.
%% @end
-spec sp_client_child_spec(atom(), parsed_connection_options(), list(pull_client_option())) -> supervisor:child_spec().
sp_client_child_spec(Name, ConnInfo, Config) ->
    ConnReg = list_to_atom(atom_to_list(Name) ++ "_conn"),
    {Name, {amqp_client_sup, start_link_sp, [Name, ConnReg, ConnInfo, Config]},
     permanent, infinity, supervisor, [amqp_client_sup]}.

%% @deprecated Please use {@link ad_client_child_spec/3} instead
client_child_spec(Name, ConnInfo, Config) ->
    ConnReg = list_to_atom(atom_to_list(Name) ++ "_conn"),
    {Name, {amqp_client_sup, start_link, [Name, ConnReg, ConnInfo, Config]},
     permanent, infinity, supervisor, [amqp_client_sup]}.

%% @doc
%% Parses the connection parameters.
%% This function takes the conneciton parameters in form of a proplist and outputs them in format used by other functions.
%% @end
-spec parse_connection_parameters(list(connection_option())) -> parsed_connection_options().
parse_connection_parameters(Props) ->
  case [proplists:get_value(E, Props)
         || E <- [host, port, username, password, virtual_host]] of
    [Host, Port, Username, Password, VHost]
      when is_list(Host),
           Port == undefined orelse is_integer(Port),
           is_binary(Username),
           is_binary(Password) ->
       VirtualHost = case VHost of
                         undefined -> <<"/">>;
                         VH -> VH
                     end,
       #amqp_params_network { username = Username, password = Password,
                              host = Host, port = Port, virtual_host = VirtualHost };
    _Otherwise ->
      exit({error, parse_connection_parameters})
  end.

File: lib/amqp_director.ex
defmodule AmqpDirector do
  @moduledoc """
  Documentation for AmqpDirector.

  This module provides wrapping for the Erlang code and is intended to be used with Elixir applications.
  """
  require AmqpDirector.Queues

  @typedoc "RabbitMQ broker connection options."
  @type connection_option ::
      {:host, String.t} |
      {:port, non_neg_integer} |
      {:username, String.t} |
      {:password, String.t} |
      {:virtual_host, String.t}

  @typedoc "The content type of the AMQP message payload."
  @type content_type :: String.t

  @typedoc "The type that a AMQP RPC Server handler function must return."
  @type handler_return_type ::  {:reply, payload :: binary, content_type} |
                                :reject |
                                :reject_no_requeue |
                                {:reject_dump_msg, String.t} |
                                :ack

  @typedoc "The handler function type for the AMQP RPC Server."
  @type handler :: (payload :: binary, content_type, type :: String.t -> handler_return_type)

  @typedoc """
  AMQP RPC Server configuration options.

  * `:consume_queue` - Specifies the name of the queue on which the server will consume messages.
  * `:consumer_tag` - Specifies the tag that the server will be identified when listening on the queue.
  * `:queue_definition` - A list of instructions for setting up the queues and exchanges. The RPC Server will call these during its
    initialization. The instructions should be created using `exchange_declare/2`, `queue_declare/2` and `queue_bind/3`. E.g.
    ```
      {:queue_definitions, [
        AmqpDirector.exchange_declare("my_exchange", type: "topic"),
        AmqpDirector.queue_declare("my_queue", exclusive: true),
        AmqpDirector.queue_bind("my_queue", "my_exchange", "some.topic.*")
      ]}
    ```
  * `:no_ack` - Specifies if the server should _NOT_ auto-acknowledge consumed messages. Defaults to `false`.
  * `:qos` - Specifies the prefetch count on the consume queue. Defaults to `2`.
  * `:reply_persistent` - Specifies the delivery mode for the AMQP replies. Setting this to `true` will make the broker log the
    messages on disk. See AMQP specification for more information. Defaults to `false`
  """
  @type server_option ::
      {:consume_queue, String.t} |
      {:consumer_tag, String.t} |
      {:queue_definitions, list(queue_definition)} |
      {:no_ack, boolean} |
      {:qos, number} |
      {:reply_persistent, boolean}

  @typedoc """
  AMQP RPC Client configuraion options.

  * `:app_id` - Specifies the identitifier of the client.
  * `:queue_definition` - A list of instructions for setting up the queues and exchanges. The RPC Client will call these during its
  initialization. The instructions should be created using `exchange_declare/2`, `queue_declare/2` and `queue_bind/3`. E.g.
  ```
    {:queue_definitions, [
      AmqpDirector.exchange_declare("my_exchange", type: "topic"),
      AmqpDirector.queue_declare("my_queue", exclusive: true),
      AmqpDirector.queue_bind("my_queue", "my_exchange", "some.topic.*")
    ]}
  ```
  * `:reply_queue` - Allows naming for the reply queue. Defaults to empty name, making the RabbitMQ broker auto-generate the name.
  * `:no_ack` - Specifies if the client should _NOT_ auto-acknowledge replies. Defaults to `false`.
  """
  @type client_option ::
      {:app_id, String.t} |
      {:queue_definitions, list(queue_definition)} |
      {:reply_queue, String.t} |
      {:no_ack, boolean}

  @typedoc """
  AMQP RPC Pull Client configuraion options.

  * `:app_id` - Specifies the identitifier of the client.
  * `:queue_definition` - A list of instructions for setting up the queues and exchanges. The RPC Client will call these during its
  initialization. The instructions should be created using `exchange_declare/2`, `queue_declare/2` and `queue_bind/3`. E.g.
  ```
    {:queue_definitions, [
      AmqpDirector.exchange_declare("my_exchange", type: "topic"),
      AmqpDirector.queue_declare("my_queue", exclusive: true),
      AmqpDirector.queue_bind("my_queue", "my_exchange", "some.topic.*")
    ]}
  ```
  """
  @type pull_client_option ::
  {:app_id, String.t} |
  {:queue_definitions, list(queue_definition)}

  @typedoc "Queue definition instructions."
  @type queue_definition :: exchange_declare | queue_declare | queue_bind

  @doc """
  Creates a child specification for an AMQP RPC server.

  This specification allows for RPC servers to be nested under any supervisor in the application using AmqpDirector. The RPC Server
  will initialize the queues it is instructed to and will then consume messages on the queue specified. The handler function will
  be called to handle each request. See `t:server_option/0` for configuration options and `t:handler/0` for the type spec
  of the handler function.

  The server handles reconnecting by itself.
  """
  @spec server_child_spec(atom, handler, list(connection_option), non_neg_integer, list(server_option)) :: Supervisor.child_spec
  def server_child_spec(name, handler, connectionInfo, count, config) do
    connectionInfo
    |> Keyword.update!(:host, &String.to_charlist/1)
    |> :amqp_director.parse_connection_parameters()
    |> (fn(connection) -> :amqp_director.server_child_spec(name, handler, connection, count, config) end).()
    |> old_spec_to_new
  end

  @doc """
  Creates a child specification for an AMQP RPC client.

  This specification allows for RPC clients to be nested under any supervisor in the application using AmqpDirector. The RPC
  client can perform queue initialization. It will also create a reply queue to consume replies on. The client can then be used
  using `AmqpDirector.Client` module API. See `t:client_option/0` for configuration options.

  The client handles reconnecting by itself.
  """
  @spec client_child_spec(atom, list(connection_option), list(client_option)) :: Supervisor.child_spec
  def client_child_spec(name,  connectionInfo, config) do
    connectionInfo
    |> Keyword.update!(:host, &String.to_charlist/1)
    |> :amqp_director.parse_connection_parameters()
    |> (fn(connection) -> :amqp_director.ad_client_child_spec(name, connection, config) end).()
    |> old_spec_to_new
  end


  @doc """
  Creates a child specification for an AMQP RPC pull client.

  This specification allows for RPC clients to be nested under any supervisor in the application using AmqpDirector. The pull client
  uses the Synchronous Pull (`#basic.get{}`) over AMQP. The client can then be used using `AmqpDirector.PullClient` module API.
  See `t:pull_client_option/0` for configuration options.

  The client handles reconnecting by itself.
  """
  @spec pull_client_child_spec(atom, list(connection_option), list(pull_client_option)) :: Supervisor.child_spec
  def pull_client_child_spec(name,  connectionInfo, config) do
    connectionInfo
    |> Keyword.update!(:host, &String.to_charlist/1)
    |> :amqp_director.parse_connection_parameters()
    |> (fn(connection) -> :amqp_director.sp_client_child_spec(name, connection, config) end).()
    |> old_spec_to_new
  end


  @typep queue_declare :: AmqpDirector.Queues.queue_declare

  @doc """
  Declares a queue on the AMQP Broker.

  This function is intended to be using within `:queue_definitions` configuration parameter of a client or a server. See
  `t:client_option/0` or `t:server_option/0` for details.

  Available options are: `:passive`, `:durable`, `:exclusive`, `:auto_delete` and `:arguments`. See AMQP specification for details on
  queue declaration.
  """
  @spec queue_declare(String.t, Keyword.t) :: queue_declare
  def queue_declare(name, params \\ []) do
    passive = Access.get(params, :passive, false)
    durable = Access.get(params, :durable, false)
    exclusive = Access.get(params, :exclusive, false)
    auto_delete = Access.get(params, :auto_delete, false)
    arguments = Access.get(params, :arguments, [])
    AmqpDirector.Queues.queue_declare(queue: name, passive: passive, durable: durable, exclusive: exclusive, auto_delete: auto_delete, arguments: arguments)
  end

  @typep queue_bind :: AmqpDirector.Queues.queue_bind

  @doc """
  Binds a queue to an exchange.

  This function is intended to be using within `:queue_definitions` configuration parameter of a client or a server. See
  `t:client_option/0` or `t:server_option/0` for details. See AMQP specification for details of queue binding.
  """
  @spec queue_bind(String.t, String.t, String.t) :: queue_bind
  def queue_bind(name, exchange, routing_key) do
    AmqpDirector.Queues.queue_bind(queue: name, exchange: exchange, routing_key: routing_key)
  end


  @typep exchange_declare :: AmqpDirector.Queues.exchange_declare

  @doc """
  Declares an exchange on the AMQP Broker.

  This function is intended to be using within `:queue_definitions` configuration parameter of a client or a server. See
  `t:client_option/0` or `t:server_option/0` for details.

  Available options are: `:passive`, `:durable`, `:auto_delete`, `:internal` and `:arguments`. See AMQP specification for details on exchange
  declaration.
  """
  @spec exchange_declare(String.t, Keyword.t) :: exchange_declare
  def exchange_declare(name, params \\ []) do
    type = Keyword.get(params, :type, "direct")
    passive = Keyword.get(params, :passive, false)
    durable = Keyword.get(params, :durable, false)
    auto_delete = Keyword.get(params, :auto_delete, false)
    internal = Keyword.get(params, :internal, false)
    arguments = Access.get(params, :arguments, [])
    AmqpDirector.Queues.exchange_declare(exchange: name, passive: passive, durable: durable, type: type, auto_delete: auto_delete, internal: internal, arguments: arguments)
  end

  defp old_spec_to_new({name, start, restart, shutdown, type, modules}), do: %{id: name, start: start, restart: restart, shutdown: shutdown, type: type, modules: modules}

end
